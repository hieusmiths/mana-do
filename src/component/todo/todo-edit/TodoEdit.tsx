import React, { useState } from 'react';

export interface AppProps {
  defaultValue: string,
  onCancel: () => void,
  onUpdate: (content: string) => void,
}

const TodoEdit = ({ defaultValue, onCancel, onUpdate }: AppProps) => {
  const [value, setValue] = useState(defaultValue);
  const onKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
    const isModify = value !== defaultValue;
    if (e.key === 'Enter') {
      if (isModify && value) {
        onUpdate(value);
      }
      onCancel();
      return;
    }
    if (e.key === 'Escape') {
      onCancel();
    }
  };
  return <input
    autoFocus
    defaultValue={defaultValue}
    onChange={(e) => setValue(e.target.value)}
    onKeyDown={onKeyDown}
    onBlur={onCancel}
  />
};

export default TodoEdit;