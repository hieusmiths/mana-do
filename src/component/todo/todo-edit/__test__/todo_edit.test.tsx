import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import TodoEdit, { AppProps } from '../TodoEdit';

describe("<TodoEdit />", () => {
  it("Should render corectly", () => {
    const defaultValue = 'YYYY';
    const onCancelMock = jest.fn();
    const onUpdateMock: jest.Mock<void, [string]> = jest.fn();
    const props: AppProps = {
      defaultValue: defaultValue,
      onCancel: onCancelMock,
      onUpdate: onUpdateMock,
    };
    const wrapper = shallow(<TodoEdit {...props} />)
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("Should render corectly default value", () => {
    const defaultValue = 'YYYY';
    const onCancelMock = jest.fn();
    const onUpdateMock: jest.Mock<void, [string]> = jest.fn();
    const props: AppProps = {
      defaultValue: defaultValue,
      onCancel: onCancelMock,
      onUpdate: onUpdateMock,
    };
    const wrapper = mount(<TodoEdit {...props} />)
    const inputValue = wrapper.find('input');
    expect(inputValue.props().defaultValue).toEqual(defaultValue);
    expect(inputValue.exists()).toBeTruthy();
  });

  it("Should onCancel when onBlur", () => {
    const defaultValue = 'YYYY';
    const onCancelMock = jest.fn();
    const onUpdateMock: jest.Mock<void, [string]> = jest.fn();
    const props: AppProps = {
      defaultValue: defaultValue,
      onCancel: onCancelMock,
      onUpdate: onUpdateMock,
    };
    const wrapper = mount(<TodoEdit {...props} />)
    const inputValue = wrapper.find('input');
    inputValue.simulate('blur');
    expect(onCancelMock).toBeCalled();
  });

  it("Should prvent onUpdate when is not modify", () => {
    const defaultValue = 'YYYY';
    const onCancelMock = jest.fn();
    const onUpdateMock: jest.Mock<void, [string]> = jest.fn();
    const props: AppProps = {
      defaultValue: defaultValue,
      onCancel: onCancelMock,
      onUpdate: onUpdateMock,
    };
    const wrapper = shallow(<TodoEdit {...props} />)
    const inputValue = wrapper.find('input');
    inputValue.simulate('change', { target: { value: defaultValue }});
    inputValue.simulate('keydown', { key: 'Enter' });
    wrapper.update();
    expect(onUpdateMock).not.toBeCalled();
  });

   it("Should onUpdate and onCancel when modify", () => {
    const defaultValue = 'YYYY';
    const onCancelMock = jest.fn();
    const onUpdateMock: jest.Mock<void, [string]> = jest.fn();
    const props: AppProps = {
      defaultValue: defaultValue,
      onCancel: onCancelMock,
      onUpdate: onUpdateMock,
    };
    const wrapper = shallow(<TodoEdit {...props} />)
    wrapper.find('input').simulate('change', { target: { value: defaultValue + 'Edit' }});
    wrapper.find('input').simulate('keyDown', { key: 'Enter' });
    wrapper.update();
    expect(onUpdateMock).toBeCalled();
    expect(onCancelMock).toBeCalled();
  });


  it("Should not run onUpdate when modify is empty", () => {
    const defaultValue = 'YYYY';
    const onCancelMock = jest.fn();
    const onUpdateMock: jest.Mock<void, [string]> = jest.fn();
    const props: AppProps = {
      defaultValue: defaultValue,
      onCancel: onCancelMock,
      onUpdate: onUpdateMock,
    };
    const wrapper = shallow(<TodoEdit {...props} />)
    wrapper.find('input').simulate('change', { target: { value: '' }});
    wrapper.find('input').simulate('keyDown', { key: 'Enter' });
    wrapper.update();
    expect(onUpdateMock).not.toBeCalled();
    expect(onCancelMock).toBeCalled();
  });

  it("Should not run onUpdate when Escape, run onCancel", () => {
    const defaultValue = 'YYYY';
    const onCancelMock = jest.fn();
    const onUpdateMock: jest.Mock<void, [string]> = jest.fn();
    const props: AppProps = {
      defaultValue: defaultValue,
      onCancel: onCancelMock,
      onUpdate: onUpdateMock,
    };
    const wrapper = shallow(<TodoEdit {...props} />)
    wrapper.find('input').simulate('keyDown', { key: 'Escape' });
    wrapper.update();
    expect(onUpdateMock).not.toBeCalled();
    expect(onCancelMock).toBeCalled();
  });

});