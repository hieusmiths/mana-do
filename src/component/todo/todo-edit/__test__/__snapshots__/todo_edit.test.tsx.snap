// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoEdit /> Should render corectly 1`] = `
<input
  autoFocus={true}
  defaultValue="YYYY"
  onBlur={[MockFunction]}
  onChange={[Function]}
  onKeyDown={[Function]}
/>
`;
